services/src/main/scala/cromwell/services/metadata/impl/MetadataStatisticsRecorder.scala

```diff
@@ -0,0 +1,116 @@
+  }
+
+
+  object MetadataStatisticsRecorderSettings {
+    val defaultCacheSize = 20000L
+    val defaultAlertInterval = 100000L
+    val defaultSubworkflowBundling = true
+
+    def apply(configSection: Option[Config]): MetadataStatisticsRecorderSettings = (configSection flatMap { conf: Config =>
+      if (conf.as[Option[Boolean]]("enabled").forall(identity)) {
+        val cacheSize: Long = conf.getOrElse("cache-size", defaultCacheSize)
+        val metadataAlertInterval: Long = conf.getOrElse("metadata-row-alert-interval", defaultAlertInterval)
+        val subworkflowBundling: Boolean = conf.getOrElse("sub-workflow-bundling", defaultSubworkflowBundling)
+        Option(MetadataStatisticsEnabled(cacheSize, metadataAlertInterval, subworkflowBundling))
+      } else None
+
+    }).getOrElse(MetadataStatisticsDisabled)
+  }
+}
+
+sealed trait MetadataStatisticsRecorder {
+  def processEvents(putEvents: Iterable[MetadataEvent]): Vector[HeavyMetadataAlert]
+}
+
+final class NoopMetadataStatisticsRecorder extends MetadataStatisticsRecorder {
+  def processEvents(putEvents: Iterable[MetadataEvent]): Vector[HeavyMetadataAlert] = Vector.empty
+}
+
+final class ActiveMetadataStatisticsRecorder(workflowCacheSize: Long = 100000L, // 100,000
+                                       metadataAlertInterval: Long = 100000L, // 100,000
+                                       bundleSubworkflowsIntoParents: Boolean = false
+                                      ) extends MetadataStatisticsRecorder {
+
+  // Statistics for each workflow
+  private val metadataWriteStatisticsCache = CacheBuilder.newBuilder()
+    .expireAfterAccess(JDuration.ofSeconds(4.hours.toSeconds))
+    .maximumSize(workflowCacheSize)
+    .build[WorkflowId, WorkflowMetadataWriteStatistics]()
+
+  def writeStatisticsLoader(workflowId: WorkflowId): Callable[WorkflowMetadataWriteStatistics] = () => WorkflowMetadataWriteStatistics(workflowId, 0L, 0L, None)
+
+  def processEvents(putEvents: Iterable[MetadataEvent]): Vector[HeavyMetadataAlert] = {
+    putEvents.groupBy(_.key.workflowId).toVector.flatMap { case (id, list) => processEventsForWorkflow(id, list)}
+  }
+
+  private def processEventsForWorkflow(workflowId: WorkflowId, events: Iterable[MetadataEvent]): Vector[HeavyMetadataAlert] = {
+    val workflowWriteStats = metadataWriteStatisticsCache.get(workflowId, writeStatisticsLoader(workflowId))
+
+    // Find a new parent record if one exists and update the statistics to record it:
+    val parentallyUpdatedStatistics = {
+      if (!bundleSubworkflowsIntoParents) workflowWriteStats
+      else if (workflowWriteStats.knownParent.isDefined) workflowWriteStats
+      else {
+        val newParentId = events.collectFirst {
+          case MetadataEvent(MetadataKey(_, None, "parentWorkflowId"), Some(MetadataValue(value, MetadataString)), _) => Try(WorkflowId(UUID.fromString(value))).toOption
+        }.flatten
+        workflowWriteStats.copy(knownParent = newParentId)
+      }
+    }
```

Comment:
Is there a reason not to get this information from the DB by refering to `rootWorkflowExecutionUuid` on `WorkflowMetadataSummaryEntries`?

Reply:
TOL: What if we record the root workflow ID instead of parent workflow ID? Won't it be easier to figure out the workflow with largest metadata events? And see easily in logs how many sub-workflows belong to a single workflow?

Reply:
Mainly because this keeps all the logic in-class and doesn't add another dependency. Also remember that the summary you're talking about is generated by summarizing the same metadata events that are already passing through here (and probably haven't passed through yet, so a negative lookup doesn't imply a lack of parent/root workflows)